File: Valid Parentheses/soln.cpp
#include<chrono>
#include<iostream>
#include<fstream>
#include<sstream>
#include<stack>
#include<string>
#include<vector>

using namespace std;

class Solution {
public:
    float isValid(string s) {
        auto t1 = std::chrono::high_resolution_clock::now();
        stack<char> Stack;
        
        for (int i = 0; i < s.size(); i++){
            string l;
            l = s[i];
            if (l == "(" || l == "{" || l == "["){
                Stack.push(s[i]);
                continue;
            }
            if (Stack.size() == 0){
                return false;
            }
            string p;
            p = Stack.top();
            Stack.pop();

            if (l == ")" && p != "("){
                return false;
            }
            if (l == "]" && p != "["){
                return false;
            }
            if (l == "}" && p != "{"){
                return false;
            }
        }
        std::chrono::duration<double, std::milli> fp_ms = std::chrono::high_resolution_clock::now() - t1;
        return fp_ms.count();
        // std::cout << "Time taken: " << fp_ms.count() << " ms, ";
        // return Stack.size() == 0;
    }
};

int main (){
    Solution soln;
	string line;
	int i = 0;
	vector<string> v;
	ifstream fin;
	fin.open ("testcase.txt");

	if (fin.is_open()){
		while (getline(fin, line)){
			stringstream ss(line);

			while (getline(ss, line, ',')){
				v.push_back(line);
			}
		}
        vector<float> times;
        string vars = v[0];
        int sample_size = 1000;
        int iters = 100;
        for (int i = 0; i < iters; i++){
            float prop = float(i) / float(iters);
            int ind = prop * vars.size();
            float md = 0;
            for (int j = 0; j < sample_size; j++){
                md += soln.isValid(vars.substr(0, ind));
            }
            times.push_back(md / sample_size);
        }
        for (int i = 0; i < times.size(); i++){
            cout << times[i] << ", ";
        }
	}
    return 0;
}

File: Valid Parentheses/soln.py
class Solution:
    def isValid(self, s: str) -> bool:
        import time

        t0 = time.time()
        
        left = {'(', '[', '{'}
        stack = []
        
        for l in s:
            if l in left:
                stack.append(l)
                continue
            if not stack:
                return False
            p = stack.pop()
            if l == ')' and p != '(':
                return False
            if l == '}' and p != '{':
                return False
            if l == ']' and p != '[':
                return False
        
        return (time.time() - t0) * 1000
        # print('{:.20f}'.format((time.time() - t0) * 1000))
        # return len(stack) == 0


with open("testcase.txt", "r") as fil:
    vars = fil.readlines()
    vars = [var.rstrip() for var in vars]
    inp = vars[0]
    times = []
    iters = 100
    sample_size = 1000
    for i in range(iters):
        prop = (i/iters)
        mid = 0
        for j in range(sample_size):
            mid += (Solution().isValid(inp[:int(prop * len(inp))]))
        times.append(round(mid / sample_size, 2))
    print(times)